torch/csrc/lazy/core/cache.h

```diff
@@ -35,6 +35,11 @@ class Cache {
   // removed from the cache.
   TypePtr Add(K key, TypePtr object) {
     std::lock_guard<std::mutex> slock(lock_);
+
+    if (!max_size_) {
```

Comment:
Why are we doing it after the lock? Is the max_size_ some thing can be changed after initialization?

Reply:
Good point. I was being too cautious since `max_size_` was not a const, even though no code actually changed it. I've pushed a change to make it const and also modified all the APIs to take this special case into account by not having to acquire the lock.